### PR TITLE
Fix when Compiling with non-standard CMake Configurations

### PR DESCRIPTION
**Pull request by mitchute**
Pull request overview
---------------------
 - Fixes when compiling with the following
   - ```EP_nocache_Psychrometrics```
   - ```EP_cache_GlycolSpecificHeat```
   - ```LINK_WITH_PYTHON=OFF```
 - Adds CMake commands and CI runs to build the following:
   - ```LINK_WITH_PYTHON=OFF```
   - ```USE_PSYCHROMETRICS_CACHING=OFF```
   - ```USE_GLYCOL_CACHING=OFF```
   - ```USE_PSYCH_STATS=ON```
   - ```USE_PSYCH_ERRORS=OFF```

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

**Comment by JasonGlazer** [comment]
@mitchute I posted #8377 which would help catch this kind of thing in the future.

**Comment by mitchute** [comment]
Thanks, @JasonGlazer. This PR is adding CI runs to the GitHub actions workflows to build with LINK_WITH_PYTHON=ON/OFF and USE_PSYCHROMETRIC_CACHING=ON/OFF, so this should address #8377. Are there other macros like this we need to catch?

**Comment by JasonGlazer** [comment]
I don't know, I presume so but I'm not sure.

**Comment by mitchute** [comment]
Here are some other macros in EnergyPlus that we may want to also consider:

- EP_Detailed_Timings
- EP_Count_Calls
- SKYLINE_MATRIX_REMOVE_ZERO_COLUMNS
- EP_cache_GlycolSpecificHeat
- GROUND_PLOT
- EP_psych_errors
- EP_psych_stats
- EP_NO_OPENGL

**Comment by Myoldmopar** [comment]
@mitchute looks like there is indeed still some issue here, but this will be a really great addition.  I don't know exactly how many of those definition conditionals we want to check...our list of possible builds could become monstrous.  We surely don't need to build special versions of these on **every single** commit:
- EP_Detailed_Timings
- EP_Count_Calls
- SKYLINE_MATRIX_REMOVE_ZERO_COLUMNS
- GROUND_PLOT

I'd be interested in what happens if we toggle the other 4.  

We could consider adding some special builds of more obscure build configurations to each tag rather than each commit.  That way CI will only do those builds when we do package building.

**Comment by mitchute** [comment]
OK, I'm not sure what the problem is currently. I haven't seen any build problems locally on Mac and Linux, but I'll investigate more.

**Comment by jasondegraw** [comment]
@myoldmopar @mitchute I'm honestly not sure if you're trolling me now, but please remove SKYLINE_MATRIX_REMOVE_ZERO_COLUMNS from the list. It isn't the flag you're looking for.

**Comment by mitchute** [comment]
> I'm honestly not sure if you're trolling me now

Wouldn't that be fun... :trollface: 

I just straight up did a search of the code for all of the preprocessor macros and listed them. No trolling intended. 😄

**Comment by mitchute** [comment]
@Myoldmopar This is looking like NREL CI may need to run CMake again to catch these new flags. If you take a look at the GitHub builds for USE_PSYCHROMETRICS_CACHING=OFF, you will see that they are failing with the same errors as the NREL CI machines which should have caching turned *on*.

**Comment by mitchute** [comment]
Currently, we have the following build matrix enabled on this branch. 

| CI Platform | Build Name             | LINK_WITH_PYTHON | USE_PSYCHROMETRICS_CACHING | USE_GLYCOL_CACHING | USE_PSYCH_STATS | USE_PSYCH_ERRORS | OPENGL_REQUIRED |
|-------------|------------------------|------------------|----------------------------|--------------------|-----------------|------------------|-----------------|
| Decent CI   | N/A  (Every push/tag)  | X                | X                          | X                  |                 | X                | X               |
| GH Actions  | python_no_psychcach    | X                |                            | X                  |                 | X                | X               |
| GH Actions  | no_python_psychcach    |                  | X                          | X                  |                 | X                | X               |
| GH Actions  | no_python_no_psychcach |                  |                            | X                  |                 | X                | X               |
| GH Actions  | no_glcol_cache         |                  | X                          |                    |                 | X                | X               |
| GH Actions  | no_opengl              |                  | X                          | X                  |                 | X                |                 |
| GH Actions  | use_psych_stats        |                  | X                          | X                  | X               | X                | X               |
| GH Actions  | no_use_psych_errors    |                  | X                          | X                  | X               |                  | X               |

I'm not sure if we want to run them all on every push, but this should at least demonstrate that they are now all working.

**Comment by mitchute** [comment]
@Myoldmopar @mjwitte this is ready for review.

File: src/EnergyPlus/Psychrometrics.hh [hunk 1]
```diff
@@ -1343,4 +1376,4 @@ namespace Psychrometrics {
 
 } // namespace EnergyPlus
 
-#endif
+#endif
```
Comment (mjwitte) [inline src/EnergyPlus/Psychrometrics.hh:1379]: This file needs an endline at the end.

File: src/EnergyPlus/Psychrometrics.hh [hunk 2]
```diff
@@ -638,17 +648,19 @@ namespace Psychrometrics {
...
 #endif
 
-    inline Real64 PsyWFnTdbH(EnergyPlusData &state, Real64 const TDB,                             // dry-bulb temperature {C}
+    inline Real64 PsyWFnTdbH([[maybe_unused]] EnergyPlusData &state,
+                             Real64 const TDB,                             // dry-bulb temperature {C}
                              Real64 const H,                               // enthalpy {J/kg}
-                             std::string const &CalledFrom = blank_string, // routine this function was called from (error messages)
-                             bool const SuppressWarnings = false           // if calling function is calculating an intermediate state
+                             [[maybe_unused]] std::string const &CalledFrom = blank_string, // routine this function was called from (error messages)
+                             [[maybe_unused]] bool const SuppressWarnings = false           // if calling function is calculating an intermediate state
```
Comment (mjwitte) [inline src/EnergyPlus/Psychrometrics.hh:663]: I was going to question adding `maybe_unused` to these arguments when they have defaults, but I see now that certain arguments do truly become unused based on the state of the `ifdef`s. That makes sense.

**Review by mjwitte — APPROVED** [review]
@mitchute The c++ changes all look ok to me. I'll yield to @Myoldmopar regarding the cmake and actions changes.

File: src/EnergyPlus/Psychrometrics.hh (hunk 1 above)
Comment (mitchute) [reply src/EnergyPlus/Psychrometrics.hh:1379]: I'll probably just add this to one of my upcoming globals branches and not spend more CI cycles on this.

Comment (Myoldmopar) [reply src/EnergyPlus/Psychrometrics.hh:1379]: :+1:

File: .github/workflows/mac_build.yml [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+
+      - name: Configure CMake
+        shell: bash
+        working-directory: ${{runner.workspace}}/EnergyPlus/build
+        run: cmake -DCMAKE_BUILD_TYPE=Release -USE_PSYCH_ERRORS=OFF ..
+
+      - name: Build EnergyPlus
+        working-directory: ${{runner.workspace}}/EnergyPlus/build
+        shell: bash
+        run: cmake --build . --target energyplus -j 2
```
Comment (Myoldmopar) [inline .github/workflows/mac_build.yml:172]: This is fantastic!

File: CMakeLists.txt [hunk 4]
```diff
@@ -392,3 +396,7 @@ mark_as_advanced(BUILD_GMOCK)
 mark_as_advanced(BUILD_GTEST)
 mark_as_advanced(gmock_build_tests)
 mark_as_advanced(BUILD_WCE_TESTING)
+mark_as_advanced(USE_PSYCHROMETRICS_CACHING)
+mark_as_advanced(USE_GLYCOL_CACHING)
+mark_as_advanced(USE_PSYCH_STATS)
+mark_as_advanced(USE_PSYCH_ERRORS)
```
Comment (Myoldmopar) [inline CMakeLists.txt:402]: It's really great having some new CMake (advanced) variables for turning compiler settings on and off.

File: src/EnergyPlus/PluginManager.cc [hunk 5]
```diff
@@ -1328,7 +1328,7 @@ namespace PluginManagement {
         }
     }
 #else
-    void PluginManager::setGlobalVariableValue([[maybe_unused]] EnergyPlusData &state, int EP_UNUSED(handle), Real64 EP_UNUSED(value))
+    void PluginManager::setGlobalVariableValue([[maybe_unused]] EnergyPlusData &state, [[maybe_unused]] int handle, [[maybe_unused]] Real64 value)
```
Comment (Myoldmopar) [inline src/EnergyPlus/PluginManager.cc:1331]: Yay for fixed syntax errors!

File: src/EnergyPlus/Psychrometrics.hh [hunk 6]
```diff
@@ -71,40 +71,41 @@ struct EnergyPlusData;
...
+    constexpr int iPsyWFnTdbRhPb = 8;
+    constexpr int iPsyPsatFnTemp = 9;
+    constexpr int iPsyTsatFnHPb = 10;
+    constexpr int iPsyTsatFnPb = 11;
+    constexpr int iPsyTsatFnPb2 = 17; // iterations
+    constexpr int iPsyRhFnTdbRhov = 12;
+    constexpr int iPsyRhFnTdbRhovLBnd0C = 13;
+    constexpr int iPsyTwbFnTdbWPb_cache = 18;
+    constexpr int iPsyPsatFnTemp_cache = 19;
+    constexpr int NumPsychMonitors = 19; // Parameterization of Number of psychrometric routines that
```
Comment (Myoldmopar) [inline src/EnergyPlus/Psychrometrics.hh:106]: Oh yes.

File: src/EnergyPlus/Psychrometrics.hh [hunk 7]
```diff
@@ -213,16 +213,16 @@ namespace Psychrometrics {
...
-    extern Array1D<cached_psat_t> cached_Psat; // DIMENSION(0:psatcache_size)
+    inline Array1D<cached_psat_t> cached_Psat; // DIMENSION(0:psatcache_size)
 #endif
 #ifdef EP_cache_PsyTsatFnPb
-    extern Array1D<cached_tsat_pb> cached_Tsat; // DIMENSION(0:tsatcache_size)
+    inline Array1D<cached_tsat_pb> cached_Tsat; // DIMENSION(0:tsatcache_size)
 #endif
 #ifdef EP_cache_PsyTsatFnHPb
-    extern Array1D<cached_tsat_h_pb> cached_Tsat_HPb; // DIMENSION(0:tsat_hbp_cache_size)
+    inline Array1D<cached_tsat_h_pb> cached_Tsat_HPb; // DIMENSION(0:tsat_hbp_cache_size)
```
Comment (Myoldmopar) [inline src/EnergyPlus/Psychrometrics.hh:225]: :+1:

**Review by Myoldmopar — APPROVED** [review]
No problems here, this will definitely keep Github Action instances busy for us :+1:

**Comment by Myoldmopar** [comment]
I thought I had merged this...well I am now.  Thanks so much for this @mitchute, this will make sure all these conditional code paths are building all the time now!

**Pull request merged**